WIP: Various updates to the Regex HOWTO (#107825)
python/cpython: Doc/howto/regex.rst

+==================+===============================================+
| ``fullmatch()``  | Determine if the RE matches the entire string |
|                  | exactly.                                      |
+------------------+-----------------------------------------------+
| ``match()``      | Determine if the RE matches at the beginning  |
|                  | of the string.                                |
+------------------+-----------------------------------------------+
| ``search()``     | Scan through a string, looking for any        |
|                  | location where the RE matches.                |
+------------------+-----------------------------------------------+
| ``findall()``    | Find all substrings where the RE matches, and |
|                  | returns them as a list.                       |
+------------------+-----------------------------------------------+
| ``finditer()``   | Returns an :term:`iterator` yielding          |
|                  | :ref:`match objects <match-objects>` for all  |
|                  | matches of the RE.                            |
+------------------+-----------------------------------------------+

> :meth:`~re.Pattern.match`, :meth:`~re.Pattern.search`, and :meth:`~re.Pattern.fullmatch` return ``None`` if no match can be found.  If

Reviewer: Split long line?
```suggestion
:meth:`~re.Pattern.match`, :meth:`~re.Pattern.search`,
and :meth:`~re.Pattern.fullmatch` return ``None`` if no match can be found.  If
```

Author: Done!

| ``search()``     | Scan through a string, looking for any        |
|                  | location where this RE matches.               |
+------------------+-----------------------------------------------+

> | ``fullmatch()``  | Determine if the RE matches the entire string |

Reviewer 2: I would move it above `match()`, because it is what people usually need when they use `match()` (and they add `$` to fix it).

Author: Done!

   >>> m.span()
   (4, 11)

The :meth:`~re.Pattern.fullmatch` method checks if the RE matches the entire
string exactly::

>    >>> p = re.compile('[a-z]+')
>    >>> p.search(' words ')
>    <re.Match object; span=(1, 6), match='words'>

Reviewer: Use the same filler text throughout the example snippet?
```suggestion
   >>> p = re.compile('[a-z]+')
   >>> p.search(' textual ')
   <re.Match object; span=(1, 6), match='textual'>
```

Author: Done!

   else:
       print('No match')

Python 3.8 added assignment expressions that shorten the above pattern
by a line::

>    p = re.compile( ... )

Reviewer 3: ```suggestion
   p = re.compile(...)
```

>    if (m := p.match( 'string goes here' )):

Reviewer 3: ```suggestion
   if m := p.match('string goes here'):
```

Author: Spaces removed, but I'd like to keep the parens.

   ['12', '11', '10']

The ``r`` prefix, making the literal a raw string literal, is needed in this
example because ``\d`` is not an escape sequence recognized in Python string literals.

> Such unrecognized sequences now produce a
> :exc:`SyntaxWarning` and will eventually become a :exc:`SyntaxError`.  See
> :ref:`the-backslash-plague`.

Reviewer: This feels redundant.
```suggestion
See :ref:`the-backslash-plague`.
```

Author: Applied; I decided to also remove the link.

You don't have to create a pattern object and call its methods; the
:mod:`re` module also provides top-level functions called :func:`~re.match`,

> :func:`~re.search`, :func:`~re.fullmatch`, :func:`~re.findall`, :func:`~re.sub`, and so forth.  These functions

Reviewer: Break long line?
```suggestion
:func:`~re.search`, :func:`~re.fullmatch`, :func:`~re.findall`,
:func:`~re.sub`, and so forth.  These functions
```

.. currentmodule:: re

Compilation flags let you modify some aspects of how regular expressions work.
They can be passed as an argument to functions such as :func:`re.compile` and
:func:`re.sub` or you can specify them in the regex pattern.

Flags are available in the :mod:`re` module under two names, a long
name such as :const:`IGNORECASE` and a short, one-letter form such as
:const:`I`. Multiple flags can be specified by bitwise OR-ing them;
``re.IGNORECASE | re.MULTILINE`` sets both the :const:`IGNORECASE` and
:const:`MULTILINE` flags, for example.

To specify them in the pattern, you can write them as an embedded
modifier at the start of the pattern that uses the short one-letter

> form: `(?i)` for a single flag or `(?mxi)` to enable multiple flags.

Reviewer 2: It is worth to mentioned "modifier spans" like `(?i:...)`. They are more powerful than global flags and modifiers.

Reviewer: I think so, but it's also okay to do that in a separate PR. We can iterate and work incrementally.

   >>> m.groups()
   ('abc', 'b')

Backreferences in a pattern allow you to specify that the contents of an
earlier capturing group must also be found at the current location in the
string. For example, ``\2`` will reference the substring matched by group 2,
succeeding only if those exact contents are found at the current position
within the string.

(Remember that Python's string literals also use a backslash followed by
numbers for including arbitrary characters in a string, so be sure to use a
raw string when incorporating backreferences in a RE.)

For example, the following RE detects doubled words in a string. ::

>    >>> p = re.compile(r'\b(\w+)\b\s+\1\b')

Reviewer 2: The second `\b` was removed intentionally. It is not needed here.

It is worth also to use possessive qualifiers here.

Reviewer: But it's fine to keep the second `\b`, and when modifying the example for some other context it might be useful. So I'd be fine with keeping it too. (Note that it's mentioned in the text below also.)

(Also, what's a possessive qualifier?)

Reviewer 2: Not exactly this example, but see the conversation in https://github.com/python/cpython/pull/21420 about redundant `\b`.

This example was fixed in https://github.com/python/cpython/pull/4443. It was incorrect without `\b` at the end, but `\b` between `\w` and `\s` is redundant by definition.

Sorry, not "possessive qualifier" but  "possessive quantifier" (although in some documents they are named "qualifiers"). A possessive quantifier is a quantifier without backtracking. It is written by adding `+` to the quantifier (as non-greed quantifiers are written by adding `?`). For example, when try to match the pattern with greedy quantifiers `\b(\w+)\s+\1\b` in "then the", a dumb backtracking engine will try to match "then then", fail, backtrack and try to match consequentially "the ", "th ", "t " until it give up. But with possessive quantifier `\b(\w++)\s++\1\b` it will not backtrack and fail quicker. It is a new feature in Python 3.11. Even if it is supported in most modern RE engines, it is relatively little known, because it was not initially supported in old RE engines.

See https://www.regular-expressions.info/possessive.html

Author: OK, I've removed the second \b and edited the text below a bit.

numbers, groups can be referenced by a name.

The syntax for a named group is one of the Python-specific extensions:
``(?P<name>...)``. Named groups behave exactly like capturing groups, and
additionally associate *name* with the group so that *name* can be used to
refer to the group in other contexts. Names should look like a Python

> identifier andonly contain letters, digits and underscores. The :ref:`match

Reviewer: ```suggestion
identifier and only contain letters, digits and underscores. The :ref:`match
```

(I don't know if these should be ASCII only, but Python identifiers nowadays can contain lots of Unicode letters -- I'm not sure to what extent the re module matches those. If the match is less than perfect this might be worth mentioning.)

Reviewer 2: The re module uses `str.isidentifier()` in string patterns. But in bytes patterns names should be ASCII only.

object <match-objects>` methods that deal with capturing groups all accept
either integers that refer to the group by number or strings that contain the
desired group's name. Named groups are still given numbers, so you can
retrieve information about a group in two ways::

   >>> p = re.compile(r'(?P<word>\b\w+\b)')

>    >>> m = p.search( '((( Lots of punctuation )))' )

Reviewer: PEP 8
```suggestion
   >>> m = p.search('((( Lots of punctuation )))')
```

           r' (?P<zonen>[-+])(?P<zoneh>[0-9][0-9])(?P<zonem>[0-9][0-9])'
           r'"')

> It's much easier to write ``m.group('zonem')``, instead of having
> to remember to retrieve group 9.

Reviewer: ```suggestion
It's much easier to write ``m.group('zonem')`` instead of having
to count groups so as to verify we must retrieve group 9.
```
